build: release: replace unmaintained actions actions-rs/toolchain, upload-release-asset.

rmstoi opened the pull request:
Replace actions: actions-rs/toolchain, upload-release-asset.

actions-rs/toolchain is unmaintained, for cross-compilation replace it
with houseabsolute/actions-rust-cross

Action upload-release-asset is unmaintained and has been archived since
Mar 4, 2021. An issue with upload-release-asset is that it creates
separate release for each job of matrix strategy. Thus, to solve these
issues, replace unmaintained upload-release-asset with upload-artifact

See: #5929

rbradford commented on .github/workflows/release.yaml:
```diff
@@ -7,43 +7,42 @@ concurrency:
         uses: actions/checkout@v4
       - name: Install musl-gcc
         run: sudo apt install -y musl-tools
+        if: contains(matrix.platform.target, 'musl')
```
nit: I think the if should come after the name.

rbradford reviewed (commented):
Thanks for doing this - did you test this on your fork?

rmstoi commented:
> Thanks for doing this - did you test this on your fork?

I tested on my fork, making a release has one [issue](https://github.com/rmstoi/cloud-hypervisor/releases) - uploaded assets  are split for each platform in matrix.  I will try to find a way to combine all assets into single release.

rmstoi commented:
Just noticed that [actions/upload-release-asset@v1](https://github.com/actions/upload-release-asset) in unmaintained, so it is a good idea to replace it with something that supports matrix build and releases. I can add it as part of this PR.

rmstoi commented:
I've updated workflow, replaced unmaintained actions/upload-release-asset@v1
with actions/upload-artifact@v3 and use softprops/action-gh-release@v1 to make release.

rmstoi commented:
I've created a test release:
https://github.com/rmstoi/cloud-hypervisor/releases/tag/v0.2

https://github.com/rmstoi/cloud-hypervisor/actions/runs/8455726793

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -7,43 +7,42 @@ concurrency:
   release:
     if: (github.event_name == 'create' && github.event.ref_type == 'tag') || github.event_name == 'merge_group'
-    name: Release
+    name: Release ${{ matrix.toolchain }} ${{ matrix.platform.target }}
```
I think it is better to keep the `${{ matrix.toolchain }}` out of the name of the workflow, so that we don't need to adjust required workflow list every time there is a toolchain change. Anyway, we won't release binaries with multiple versions of toolchain.

rmstoi replied:
done

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -7,43 +7,42 @@ concurrency:
-    name: Release
+    name: Release ${{ matrix.toolchain }} ${{ matrix.platform.target }}
+    strategy:
+      fail-fast: false
```
Why do we need to disable fail-fast?

rmstoi replied:
Setting fail-fast to false prevents GitHub from cancelling all in-progress jobs if any matrix job fails. This is useful in case build on one of the matrix platforms fails, GitHub will still continue builds on other platforms.

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -7,43 +7,42 @@ concurrency:
+        - target: aarch64-unknown-linux-musl
+          command: build
+          args:  --all --release
+        toolchain:
+          - 1.74
+          #- stable
+          #- beta
+          #- nightly
```
We should also remove the `toolchain` from the matrix list, and directly specify toolchain version say `1.74` for now in the steps, or make it as a part of the `matrix.platform` if you prefer to have it in a central place.

rmstoi replied:
done, removed

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -56,7 +55,9 @@ jobs:
           prerelease: true
       - name: Upload cloud-hypervisor
-        if: github.event_name == 'create' && github.event.ref_type == 'tag'
+        if: |
```
Can you share a reference to explain why would a prefix of `|` be needed here?

rmstoi replied:
It is used in workflow `if` to write multi-line if conditions. It improves readability to split conditions to multiple lines instead of having one long line. More info [here](https://github.com/orgs/community/discussions/25641).

According to [YAML spec](https://yaml.org/spec/1.2.2/#81-block-scalar-styles) `|` marks literal block scalar header.

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -56,7 +55,9 @@ jobs:
-        if: github.event_name == 'create' && github.event.ref_type == 'tag'
+        if: |
+          startsWith(matrix.platform.target, 'x86_64-unknown-linux-gnu') &&
+          github.event_name == 'create' && github.event.ref_type == 'tag'
```
I believe a cleaner syntax would be:

`if: ${{ github.event_name == 'create' && github.event.ref_type == 'tag' && matrix.platform.target = 'x86_64-unknown-linux-gnu'}}`

rmstoi replied:
done

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -43,95 +57,19 @@ jobs:
-          asset_content_type: application/octet-stream
+          name: Artifacts for ${{ matrix.platform.target }}
+          path: |
+            cloud-hypervisor*
+            ch-remote*
```
Why not using `matrix.platform.name_ch` and `matrix.platform.name_ch_remote`?

rmstoi replied:
Good idea, done

likebreath reviewed (changes requested):
@rmstoi Thank you for the contribution. Some comments and questions below.

rbradford commented:
Hi @rmstoi - I think the last commit should be squashed into the earlier ones?

rmstoi commented:
> Hi @rmstoi - I think the last commit should be squashed into the earlier ones?

Correct, that was in my intentions, needed first to resolve all feedback from @likebreath. I've now squashed changes to one commit.

rmstoi commented:
Latest release on my fork is [here](https://github.com/rmstoi/cloud-hypervisor/releases/tag/v0.4)

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -3,152 +3,94 @@ on: [create, merge_group]
+      matrix:
+        platform:
+        - target: x86_64-unknown-linux-gnu
+          command: build
```
nit: the `command` option does not change and does not need to be a part of the matrix.

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -3,152 +3,94 @@ on: [create, merge_group]
       - name: Install musl-gcc
+        if: contains(matrix.platform.target, 'musl')
         run: sudo apt install -y musl-tools
       - name: Create release directory
+        if: |
+          github.event_name == 'create' && github.event.ref_type == 'tag' &&
+          matrix.platform.target == 'x86_64-unknown-linux-gnu'
         run: rsync -rv --exclude=.git . ../cloud-hypervisor-${{ github.event.ref }}
```
I believe this step can be moved later, say right before the `Vendor` step.

rmstoi replied:
I think this should be done before build, as it previously was, otherwise build artifacts will be included in vendored source archive.

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -3,152 +3,94 @@ on: [create, merge_group]
-          asset_name: cloud-hypervisor-static-aarch64
-          asset_content_type: application/octet-stream
-      - name: Upload static AArch64 ch-remote
+          command: ${{ matrix.platform.command }}
```
Ditto. This is simpler:

 `command: build`

likebreath commented on .github/workflows/release.yaml:
```diff
@@ -3,152 +3,94 @@ on: [create, merge_group]
+        run: |
+          cp target/${{ matrix.platform.target }}/release/cloud-hypervisor ./${{ matrix.platform.name_ch }}
+          cp target/${{ matrix.platform.target }}/release/ch-remote ./${{ matrix.platform.name_ch_remote }}
+      - name: Upload Release Artifacts
+        uses: actions/upload-artifact@v3
```
I believe this step needs to be behind condition: `if: github.event_name == 'create' && github.event.ref_type == 'tag'`

likebreath reviewed (commented):
Looking good in general. Some more comments below.

rmstoi commented:
Latest release on my fork is [here](https://github.com/rmstoi/cloud-hypervisor/releases/tag/v0.5)
